### PR TITLE
MeshClass

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/w3d/renderer/mesh.cpp [hunk 1]
```diff
@@ -14,21 +14,765 @@
...
+    captainslog_assert(m_model);
+    return m_model->Intersect_OBBox(local_test);
+}
+
+void MeshClass::Get_Obj_Space_Bounding_Sphere(SphereClass &sphere) const
+{
+    if (m_model) {
+        m_model->Get_Bounding_Sphere(&sphere);
+    } else {
+        sphere.Center.Set(0, 0, 0);
```
Comment (tomsons26) [inline src/w3d/renderer/mesh.cpp]: Use Init here, i want to phase out this way of setting sphere

Comment (jonwil) [reply src/w3d/renderer/mesh.cpp]: Fixed.

File: src/w3d/renderer/mesh.cpp [hunk 2]
```diff
@@ -14,21 +14,764 @@
...
+    Vector3 sc;
+    sc.X = sc.Y = sc.Z = scale;
+    Make_Unique(false);
+    m_model->Make_Geometry_Unique();
+    m_model->Scale(sc);
+
+    Invalidate_Cached_Bounding_Volumes();
+    RenderObjClass *container = Get_Container();
+
+    if (container) {
```
Comment (duncanspumpkin) [inline src/w3d/renderer/mesh.cpp]: != nullptr

File: src/w3d/renderer/mesh.cpp [hunk 3]
```diff
@@ -14,21 +14,764 @@
...
+    sc.Y = scaley;
+    sc.Z = scalez;
+    Make_Unique(false);
+    m_model->Make_Geometry_Unique();
+    m_model->Scale(sc);
+
+    Invalidate_Cached_Bounding_Volumes();
+    RenderObjClass *container = Get_Container();
+
+    if (container) {
```
Comment (duncanspumpkin) [inline src/w3d/renderer/mesh.cpp]: != nullptr

File: src/w3d/renderer/mesh.cpp [hunk 4]
```diff
@@ -14,21 +14,764 @@
...
+            opacity = -1.0f;
+        } else {
+            opacity = pass->Peek_Material()->Get_Opacity();
+            pass->Peek_Material()->Set_Opacity(m_opacityOverride);
+        }
+
+        if (m_emissiveScale != 1.0f) {
+            VertexMaterialClass *m = pass->Get_Material();
+
+            if (m) {
```
Comment (duncanspumpkin) [inline src/w3d/renderer/mesh.cpp]: != nullptr

File: src/w3d/renderer/mesh.cpp [hunk 5]
```diff
@@ -14,21 +14,764 @@
...
+    }
+
+    return count;
+}
+
+PrototypeClass *MeshLoaderClass::Load_W3D(ChunkLoadClass &cload)
+{
+    MeshClass *mesh = new MeshClass();
+
+    if (mesh->Load_W3D(cload)) {
```
Comment (duncanspumpkin) [inline src/w3d/renderer/mesh.cpp]: != W3D_ERROR_OK

File: src/w3d/renderer/mesh.h [hunk 6]
```diff
@@ -118,3 +128,31 @@ class MeshClass : public W3DMPO, public RenderObjClass
...
+
+private:
+    RenderObjClass *m_proto;
+};
+
+class MeshLoaderClass : public PrototypeLoaderClass
+{
+public:
+    virtual int Chunk_Type() { return W3D_CHUNK_MESH; }
+    virtual PrototypeClass *Load_W3D(ChunkLoadClass &cload);
```
Comment (duncanspumpkin) [inline src/w3d/renderer/mesh.h]: Think these are override s

**Review by duncanspumpkin — COMMENTED** [review]
There are a few other != nullptr in here as well but I've only marked the first few

File: src/w3d/renderer/coltest.h [hunk 7]
```diff
@@ -52,8 +52,8 @@ class RayCollisionTestClass : public CollisionTestClass
     RayCollisionTestClass(const LineSegClass &ray,
         CastResultStruct *res,
         int collision_type = COLLISION_TYPE_0,
-        bool bool1 = false,
-        bool bool2 = false);
+        bool ignore_translucent_meshes = false,
```
Comment (tomsons26) [inline src/w3d/renderer/coltest.h]: wrong names

File: src/w3d/renderer/mesh.cpp [hunk 8]
```diff
@@ -14,21 +14,764 @@
...
+        int num_passes = m_model->Get_Pass_Count();
+        captainslog_assert(num_passes > 0);
+        int poly_count = m_model->Get_Polygon_Count();
+        return num_passes * poly_count;
+    } else {
+        return 0;
+    }
+}
+
+void MeshClass::Render(RenderInfoClass &rinfo)
```
Comment (tomsons26) [inline src/w3d/renderer/mesh.cpp:231]: add asserts from wb

File: src/w3d/renderer/mesh.cpp [hunk 9]
```diff
@@ -14,21 +14,764 @@
...
+            bool render_base_passes =
+                ((rinfo.Current_Override_Flags() & RenderInfoClass::RINFO_OVERRIDE_ADDITIONAL_PASSES_ONLY) == 0);
+            bool is_alpha = (m_model->Get_Single_Shader().Get_Alpha_Test() == ShaderClass::ALPHATEST_ENABLE)
+                || (m_model->Get_Single_Shader().Get_Src_Blend_Func() == ShaderClass::SRCBLEND_SRC_ALPHA);
+
+            if ((rinfo.Current_Override_Flags() & RenderInfoClass::RINFO_OVERRIDE_SHADOW_RENDERING) && (is_alpha == true)) {
+                render_base_passes = true;
+            }
+
+            if (render_base_passes) {
```
Comment (tomsons26) [inline src/w3d/renderer/mesh.cpp:276]: conditional on `  if ( *RenderInfoClass::Current_Override_Flags(rinfo) & 8 && render_base_passes == 1 || v15 )`
v15 being    v15 = ~(v13 >> 2) & 1;

File: src/w3d/renderer/mesh.cpp [hunk 10]
```diff
@@ -14,21 +14,764 @@
...
+    }
+}
+
+void MeshClass::Generate_Culling_Tree()
+{
+    m_model->Generate_Culling_Tree();
+}
+
+void MeshClass::Add_Dependencies_To_List(DynamicVectorClass<StringClass> &file_list, bool textures_only)
+{
```
Comment (tomsons26) [inline src/w3d/renderer/mesh.cpp:713]: add the asserts

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**